src/authentication/file-based-authentication-handler.js

```diff
@@ -0,0 +1,208 @@
+'use strict';
+
+const crypto = require( 'crypto' );
+const JsonLoader = require( '../utils/json-loader' );
+const utils = require( '../utils/utils' );
+const EventEmitter = require( 'events' ).EventEmitter;
+const UNDEFINED = 'undefined';
+const STRING = 'string';
+const STRING_CHARSET = 'base64';
+
+/**
+ * This authentication handler reads a list of users and their associated password (either
+ * hashed or in cleartext ) from a json file. This can be useful to authenticate smaller amounts
+ * of clients with static credentials, e.g. backend provider that write to publicly readable records
+ *
+ * @public
+ * @extends {EventEmitter}
+ */
+module.exports = class FileBasedAuthenticationHandler extends EventEmitter{
+
+	/**
+	 * Creates the class, reads and validates the users.json file
+	 *
+	 * @param   {Object} settings
+	 * @param   {String} settings.path path to the user file
+	 * @param   {String} settings.hashAlgo the name of a HMAC digest algorithm, a.g. 'sha512'
+	 * @param   {Int} settings.iterations the amount of times the algorithm should be applied
+	 * @param   {Int} settings.keyLength the length of the resulting key
+	 *
+	 * @constructor
+	 * @returns {void}
+	 */
+	constructor( settings ) {
+		super();
+		this.isReady = false;
+		this._jsonLoader = new JsonLoader();
+		this._validateSettings( settings );
+		this._settings = settings;
+		this._base64KeyLength = 4 * Math.ceil( this._settings.keyLength / 3 );
+		this._jsonLoader.load( settings.path, this._onFileLoad.bind( this ) );
+	}
+
+	/**
+	 * Main interface. Authenticates incoming connections
+	 *
+	 * @param   {Object}   connectionData
+	 * @param   {Object}   authData
+	 * @param   {Function} callback
+	 *
+	 * @public
+	 * @implements {PermissionHandler.isValidUser}
+	 * @returns {void}
+	 */
+	isValidUser( connectionData, authData, callback ) {
+		if( typeof authData.username !== STRING ) {
+			callback( 'missing authentication parameter username', false );
+			return;
+		}
+
+		if( typeof authData.password !== STRING ) {
+			callback( 'missing authentication parameter password', false );
+			return;
+		}
+
+		var userData = this._data[ authData.username ];
+
+		if( !userData ) {
+			callback( null, false, null );
+			return;
+		}
+
+		if( this._settings.hash ) {
+			this._isValid( authData.password, userData.password, userData.data, callback );
```

Comment:
This is probably the only thing I think may be wierd, also might be good to have a json file with:

{ 
username: string
password: hash || string
data: userdata
}